### DISCUSSION TITLE
How to delete existing history of node when restarting server?

### DISCUSSION THREAD

**Discussion by andreasKarle** [FreeOpcUa/python-opcua#1227] [Q&A] [answered]
Hi,

first of all thanks for the great work here!

When i start a local server, add a node and add sqlite to historize the data and then restart this server and read the history with read_raw_history() i can get more values than i added after restarting the server.
The states are all good and history values are different from 0. So i don't know which data is since the restart.
Seems like the values i added before the restart are still accessable.

I could set the period argument but is there a way of clearing the sql data?
Obviously this should be easily done by interacting with sqlite but is there a history_manager method?
(sorry i have no experience with sql)

I tried this with example: python-opcua/examples/server-datavalue-history.py

**Comment by AndreasHeine** [comment, marked as answer]
what is your history manger config?
HistoryDict or HistorySQLite?

**Reply by AndreasHeine** [reply]
>     server.iserver.history_manager.set_storage(HistorySQLite("my_datavalue_history.sql"))

it is stored in a file (persistent)

> server.iserver.history_manager.set_storage(HistorySQLite(":memory:"))

its stored in memory (not persistent)

**Reply by andreasKarle** [reply]
its in the file:

> server.iserver.history_manager.set_storage(HistorySQLite("my_datavalue_history.sql"))

I already solved it by deleting the table but using a protected member _conn from server.iserver.history_manager.storage ...

`    
node_id = myvar.nodeid\
table = str(node_id.NamespaceIndex) + '_' + str(node_id.Identifier)\
query = 'DELETE FROM "{tn}"'.format(tn=table)\
server.iserver.history_manager.storage._conn.cursor().execute(query)`

**Reply by andreasKarle** [reply]
ah so i only need to use the inpersistent database.
that makes sense.

thanks for the fast reply

**Reply by AndreasHeine** [reply]
my recommendation: if you want no persistence while developing -> just use ":memory:" instead of "my_datavalue_history.sql"

**Reply by zerox1212** [reply]
FYI memory dict will probably grow forever until python crashes. If I were you I would just use a file operation to delete the SQL file before your OPC UA stuff starts.

**Reply by andreasKarle** [reply]
Thats actually a good point because i am loading a lot of nodes from an actual machine operation from a csv file to the server and this might consume a lot of memory fast.

Thanks!

But i could also prevent the crash by setting the count argument to limit the max number of values when i connect the variable to the history, right?
`    server.historize_node_data_change(var1, period=None, count=5000)
`
As long as i dont have to many nodes...

**Reply by zerox1212** [reply]
Sure, I would test it though. I don't remember if the history interface respects count limit. It's been a long time.
